File: backend/core-api/src/main/java/io/basearchitecture/core/client/ai/AiServiceClientException.java
package io.basearchitecture.core.client.ai;

public final class AiServiceClientException extends RuntimeException {

    public enum Failure {
        UPSTREAM_ERROR,
        INVALID_RESPONSE,
        UNAVAILABLE,
        TIMEOUT
    }

    private final Failure failure;

    private AiServiceClientException(Failure failure, String message, Throwable cause) {
        super(message, cause);
        this.failure = failure;
    }

    public static AiServiceClientException upstreamError(String message, Throwable cause) {
        return new AiServiceClientException(Failure.UPSTREAM_ERROR, message, cause);
    }

    public static AiServiceClientException invalidResponse(String message, Throwable cause) {
        return new AiServiceClientException(Failure.INVALID_RESPONSE, message, cause);
    }

    public static AiServiceClientException unavailable(Throwable cause) {
        return new AiServiceClientException(
                Failure.UNAVAILABLE,
                "AI Service could not be reached",
                cause);
    }

    public static AiServiceClientException timeout(Throwable cause) {
        return new AiServiceClientException(
                Failure.TIMEOUT,
                "AI Service request timed out",
                cause);
    }

    public Failure failure() {
        return failure;
    }
}

File: backend/core-api/src/main/kotlin/ai/govbiz/core/client/ai/AiServiceClientException.kt
package ai.govbiz.core.client.ai

class AiServiceClientException private constructor(
    val failure: Failure,
    message: String,
    cause: Throwable?,
) : RuntimeException(message, cause) {

    enum class Failure {
        UPSTREAM_ERROR,
        INVALID_RESPONSE,
        UNAVAILABLE,
        TIMEOUT,
    }

    companion object {
        fun upstreamError(message: String, cause: Throwable?): AiServiceClientException =
            AiServiceClientException(Failure.UPSTREAM_ERROR, message, cause)

        fun invalidResponse(message: String, cause: Throwable?): AiServiceClientException =
            AiServiceClientException(Failure.INVALID_RESPONSE, message, cause)

        fun unavailable(cause: Throwable?): AiServiceClientException =
            AiServiceClientException(
                Failure.UNAVAILABLE,
                "AI Service could not be reached",
                cause,
            )

        fun timeout(cause: Throwable?): AiServiceClientException =
            AiServiceClientException(
                Failure.TIMEOUT,
                "AI Service request timed out",
                cause,
            )
    }
}
